### DISCUSSION TITLE
Reading pending atom state in write operation. We throw a promise for now

### DISCUSSION THREAD

**Discussion by lili21** [pmndrs/jotai#1143] [Q&A] [answered]
I'm trying to manually set async atom, and it throw the `Reading pending atom state in write operation. We throw a promise for now` message. 

How should I handle that ? 

```javascript
const replyCommentListAtom = atomFamily(comment_id => atomWithDefault(async () => {
  const result = await fakeFetchReplyCommentList(comment_id)
  return result.data
}

const postReplyComment  = atom(null, async (get, set, comment_id) => {
  const result = await fakeCreateReplyComment(comment_id)

  set(replyCommentListAtom(comment_id), prev => {
    return [
      result.data,
      ...prev,
    ]
  })
})
```

if I never get the replyCommentList value, and set it, it will throw the error.

**Comment by dai-shi** [comment, marked as answer]
Thanks for the real use case. That message is something we are unsure how to deal with.
In your case, you want to avoid reading async atom in write.
```js
const baseReplyCommentListAtom = atomFamily(() => atom(null));

const replyCommentListAtom = atomFamily((comment_id) => atom(async (get) => {
  const overwritten = get(baseReplyCommentListAtom(commnet_id));
  if (overwritten) return overwritten;
  const result = await fakeFetchReplyCommentList(comment_id);
  return result.data;
}));

const postReplyComment  = atom(null, async (get, set, comment_id) => {
  const result = await fakeCreateReplyComment(comment_id);
  const prev = get(baseReplyCommnetListAtom(comment_id));
  set(baseReplyCommentListAtom(comment_id), [
    result.data,
    ...(prev || []),
  ]);
});
```

However, if this will throw away the first result. If it's a problem in your case, we need to do something different. Let me know in such a case.

**Reply by lili21** [reply]
the replyCommentList may not be empty when I manually set it.  in such case,  the baseReplyCommentList need to be set to the return value of `fakeFetchReplyCommentList`

**Reply by dai-shi** [reply]
> the replyCommentList may not be empty when I manually set it. in such case, the baseReplyCommentList need to be set to the return value of `fakeFetchReplyCommentList`

You probably mean `replayCommentListAtom` needs to be writable. How's this?

```js
const baseReplyCommentListAtom = atomFamily(() => atom(null));

const replyCommentListAtom = atomFamily((comment_id) => atom(async (get) => {
  const overwritten = get(baseReplyCommentListAtom(commnet_id));
  if (overwritten) return overwritten;
  const result = await fakeFetchReplyCommentList(comment_id);
  return result.data;
}, (get, set, update) => {
  set(baseReplyCommentListAtom(commnet_id), update)
}));

const postReplyComment  = atom(null, async (get, set, comment_id) => {
  const result = await fakeCreateReplyComment(comment_id);
  const prev = get(baseReplyCommnetListAtom(comment_id));
  set(baseReplyCommentListAtom(comment_id), [
    result.data,
    ...(prev || []),
  ]);
});
```

**Reply by lili21** [reply]
what I'm building is something like this
![image](https://user-images.githubusercontent.com/4456380/167260900-43a580dc-d18f-457c-9a07-0b344c718c87.png)

user can reply the comment, which will update the replyCommentList.
each comment has a replyCommentList, it may contains some data, and it may be empty. and the comment contains the count of its replies. 

so the logic here is , if the count is 0, I won't get the `replyCommentListAtom`, otherwise I will get it - which will trigger the fetch.

hope it be more clear now.

**Reply by dai-shi** [reply]
how would `postReplyComment` work? Especially, `[result.data, ...prev]` part.

**Reply by lili21** [reply]
`postReplyComment` will create a reply for a comment,  which will update its `replyCommentList` - add the new reply to the top of the list

**Reply by dai-shi** [reply]
If I understand it right, this is what I have.
```js
const baseReplyCommentListAtom = atomFamily(() => atom(null));

const replyCommentListAtom = atomFamily((comment_id) => atom(async (get) => {
  const overwritten = get(baseReplyCommentListAtom(commnet_id));
  if (overwritten) return overwritten;
  const result = await fakeFetchReplyCommentList(comment_id);
  return result.data;
}, (get, set, update) => {
  set(baseReplyCommentListAtom(commnet_id), update)
}));

const postReplyComment  = atom(null, async (get, set, { comment_id, new_comment }) => {
  let prev = get(baseReplyCommnetListAtom(comment_id));
  if (!prev) {
    prev = (await fakeFetchReplyCommentList(comment_id)).data;
  }
  const result = await fakeCreateReplyComment(comment_id, new_comment);
  set(baseReplyCommentListAtom(comment_id), [
    result.data,
    ...prev),
  ]);
});
```

**Reply by lili21** [reply]
in this way, `fakeFetchReplyCommentList` would run twice isn't it ? 
the first one is when I get the `replyCommentListAtom` and render it 
the second one is when I run the `postReplyComment`

**Reply by dai-shi** [reply]
Yes, to avoid that, you need to reach out the undocumented/unstable behavior.

```js
const postReplyComment  = atom(null, async (get, set, { comment_id, new_comment }) => {
  const prev = await get(baseReplyCommnetListAtom(comment_id), { unstable_promise: true });
  const result = await fakeCreateReplyComment(comment_id, new_comment);
  set(baseReplyCommentListAtom(comment_id), [
    result.data,
    ...prev),
  ]);
});
```

But, I guess if we want to avoid running fetch twice, we could cache it outside jotai.

(Or, hack around with `onMount`. There can be various ways to work around this. The question is what's the best practice.)
